include/nbl/video/SPhysicalDeviceLimits.h

```diff
@@ -0,0 +1,497 @@
+    //uint32_t                             maxUpdateAfterBindDescriptorsInAllPools;
+    //VkBool32                             shaderUniformBufferArrayNonUniformIndexingNative;
+    //VkBool32                             shaderSampledImageArrayNonUniformIndexingNative;
+    //VkBool32                             shaderStorageBufferArrayNonUniformIndexingNative;
+    //VkBool32                             shaderStorageImageArrayNonUniformIndexingNative;
+    //VkBool32                             shaderInputAttachmentArrayNonUniformIndexingNative;
+    //VkBool32                             robustBufferAccessUpdateAfterBind;
+    //VkBool32                             quadDivergentImplicitLod;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindSamplers;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindUniformBuffers;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindStorageBuffers;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindSampledImages;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindStorageImages;
+    //uint32_t                             maxPerStageDescriptorUpdateAfterBindInputAttachments;
+    //uint32_t                             maxPerStageUpdateAfterBindResources;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindSamplers;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindUniformBuffers;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindUniformBuffersDynamic;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindStorageBuffers;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindStorageBuffersDynamic;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindSampledImages;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindStorageImages;
+    //uint32_t                             maxDescriptorSetUpdateAfterBindInputAttachments;
+    //VkResolveModeFlags                   supportedDepthResolveModes;
+    //VkResolveModeFlags                   supportedStencilResolveModes;
+    //VkBool32                             independentResolveNone;
+    //VkBool32                             independentResolve;
+    //VkBool32                             filterMinmaxSingleComponentFormats;
+    //VkBool32                             filterMinmaxImageComponentMapping;
+    //uint64_t                             maxTimelineSemaphoreValueDifference;
+    //VkSampleCountFlags                   framebufferIntegerColorSampleCounts;
+            
+    /* Vulkan 1.3 Core  */
+    //uint32_t              minSubgroupSize;
+    //uint32_t              maxSubgroupSize;
+    //uint32_t              maxComputeWorkgroupSubgroups;
+    //VkShaderStageFlags    requiredSubgroupSizeStages;
```

Comment:
I think we want to expose these as they are related to subgroup sizes

Reply:
we do, subgroups very useful we want to use them for the scan, etc.

Reply:
As for GL

## Min & Max Subgroup Sizes

### Nvidia

Min and Max are always 32

### AMD

GCN (including Navi) 64

RDNA onwards 32

### Intel

I think min is 4 and max is 16, but ask @nanokatze on GP discord

### Others

min 4 and max 64 (or even 128  if you can find a vulkan report to back that number up)

## maxComputeWorkgroupSubgroups

Divide max workgroup size by min subgroup size ;)

Reply:
Intel is min 8 and max 32

Reply:
> Intel is min 8 and max 32

even in non-compute stages?

Reply:
Yes: narrower subgroup allows for more subgroups to be in-flight at once, same as trading off no. of accessible registers for more subgroups on other hardware

Reply:
How can I differtiate between GCN and RDNA?
I'll just set max to 32 when AMD

Reply:
> How can I differtiate between GCN and RDNA? I'll just set max to 32 when AMD

Yeah its not easy to differentiate GCN and RDNA except by `GL_RENDERER`

## if nothing is available

Nvidia: min and max is 32
AMD: min 32 max 64
Intel: min 8 max 32
Others: min 4 max 64

## if KHR_shader_subgroup is available and `SUBGROUP_FEATURE_BASIC_BIT_KHR` feature bit present

Just query the stuff from https://www.khronos.org/registry/OpenGL/extensions/KHR/KHR_shader_subgroup.txt

Namely `SUBGROUP_SIZE_KHR` is a maximum value you can set reliably, minimum you can't be sure about but:

### Nvidia is 32 minimum

### AMD is 32 minimum
Without bothering to do GCN vs RDNA detection

### Intel is 8 minimum

### Others are 4

## If `ARB_shader_ballot` is available (desktop GL only)

We can give this to @Hazardu, to be done WITHOUT Nabla API (plain GL), because this would be done during the temporary context creation, ergo no queues, etc..

Run a dummy vertex,geometry,tessellation,fragment pipeline (reasonable workload, few thousand verts, render into the default back buffer/FBO), have them write `gl_SubGroupInvocationARB` with `atomicMin` and `atomicMax` into an SSBO so we can get ranges.

Also run 2 dummy compute shaders:
- 16x16x1 workgroup and a 128x64x1 dispatch grid
- 256x1x1 workgroup and 16k x 1 x 1 dispatch grid
and do the same.

Reply:
@Erfan-Ahmadi lets assign the `ARB_shader_ballot` runtime test to @deprilula28 ?

Reply:
yep

Reply:
moved over the description of ARB_shader_ballot to a todo